netket/graph/custom_graph.py

```diff
@@ -0,0 +1,99 @@
+from abstract_graph import PyAbstractGraph
+
+import numpy as _np
+import networkx as _nx
+
+class PyCustomGraph(PyAbstractGraph):
+	r"""A custom graph, specified by a list of edges and optionally colors."""
+
+	def __init__(self, edges=[]):
+		"""
+		Constructs a new graph given a list of edges.
+
+		Args:
+		    edges: If `edges` has elements of type `Tuple[int, int]` it is treated
+		        as a list of edges. Then each element `(i, j)` means a connection
+		        between sites `i` and `j`. It is assumed that `0 <= i <= j`. Also,
+		        `edges` should contain no duplicates. If `edges` has elements of
+		        type `Tuple[int, int, int]` each element `(i, j, c)` represents an
+		        edge between sites `i` and `j` colored into `c`. It is again assumed
+		        that `0 <= i <= j` and that there are no duplicate elements in `edges`.
+
+
+		Examples:
+		    A 10-site one-dimensional lattice with periodic boundary conditions can be
+		    constructed specifying the edges as follows:
+
+		    >>> import netket
+		    >>> g=netket.graph.PyCustomGraph([[i, (i + 1) % 10] for i in range(10)])
+		    >>> print(g.n_sites)
+		    10
+		"""
+
+		assert isinstance(edges, list)
+		type_condition = [isinstance(edge, list) or isinstance(edge, tuple) for edge in edges]
+		if False in type_condition:
+			raise ValueError("edges must be a list of lists or tuples")
+
+		edges_array = _np.array(edges, dtype=_np.int32)
+		if edges_array.ndim != 2:
+			raise ValueError("edges must be a list of lists or tuples of the same length (2 or 3)")
+
+		if not(edges_array.shape[1] == 2 or edges_array.shape[1] == 3):
+			raise ValueError("edges must be a list of lists or tuples of the same length (2 or 3), where the third column indicates the color")
+
+		self.graph = _nx.MultiGraph()
```

Comment:
with MultiGraph and not Graph? in the c++ case we don't allow duplicate edges

Reply:
Because of the colours. Maybe one could have the case where the edges (0,1,0) and (0,1,1) exist because one operator acts on the 0 colour, whereas another operator acts on the 1 colour.